mlflow/utils/databricks_utils.py

```diff
@@ -469,6 +486,67 @@ def get_databricks_host_creds(server_uri=None):
     _fail_malformed_databricks_auth(profile)
 
 
+def _model_serving_env_databricks_host_creds():
+    # Since we do not record OAuth expiration time in OAuth file, perform periodic refresh
+    # of OAuth environment variable cache here. As currently configured (02/24) OAuth token
+    # in model serving environment is guaranteed to have at least 30 min remaining on TTL
+    # at any point in time but refresh at higher rate of every 10 min here to be safe
+    # and in case those values change in the future.
+    OAUTH_CACHE_REFRESH_DURATION_SEC = 10 * 60
+    OAUTH_CACHE_ENV_VAR = "DATABRICKS_DEPENDENCY_OAUTH_CACHE"
+    OAUTH_CACHE_EXPIRATION_ENV_VAR = "DATABRICKS_DEPENDENCY_OAUTH_CACHE_EXIRY_TS"
+    MODEL_SERVING_HOST_ENV_VAR = "DATABRICKS_MODEL_SERVING_HOST_URL"
+
+    # check if dependency is cached in env var before reading from file
+    oauth_token = ""
+    if (
+        OAUTH_CACHE_ENV_VAR in os.environ
+        and OAUTH_CACHE_EXPIRATION_ENV_VAR in os.environ
+        and float(os.environ[OAUTH_CACHE_EXPIRATION_ENV_VAR]) > time.time()
+    ):
+        oauth_token = os.environ[OAUTH_CACHE_ENV_VAR]
+    else:
+        oauth_token = _get_model_dependency_oauth_token()
+        os.environ[OAUTH_CACHE_ENV_VAR] = oauth_token
+        os.environ[OAUTH_CACHE_EXPIRATION_ENV_VAR] = str(
+            time.time() + OAUTH_CACHE_REFRESH_DURATION_SEC
```

Comment:
Q:

Assuming the token file is refreshed per 30min, and it was refreshed at time A, then 25min passed, then L509 is invoked at time (A + 25min), then it got a token with expiration time 10min, then at time (A + 32min) it still treats the token as valid. Shall we address the issue or anything I misunderstood ?

Reply:
Ahh so basically the token is guaranteed to have at least 30 min TTL remaining when invoked. The full life of a token is 60 minutes and token gets refreshed every 30 min, so in the worse case we read token right before new one comes in which case there is still at least 30 min remaining on the TTL. I think the 10 min refresh is just to be extra cautious or in case these values change in the future, but in reality the refresh could be as high as 29 min 59 seconds